[hip] Enable caching in the hip async allocator

AWoloszyn opened the pull request.

AWoloszyn committed:
[wip][hip] Enable caching in the hip async allocator

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -15,11 +15,28 @@
 static const char* IREE_HAL_HIP_ALLOCATOR_ID = "HIP unpooled";
 #endif  // IREE_TRACING_FEATURE_ALLOCATION_TRACKING
 
+#define iree_hal_hip_async_allocation_default_queue_size 8
```
```suggestion
#define IREE_HAL_HIP_ASYNC_ALLOCATION_DEFAULT_QUEUE_SIZE 8
```
style: all caps for macro literals
https://google.github.io/styleguide/cppguide.html#Macro_Names

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -132,6 +183,7 @@ static iree_allocator_t iree_hal_hip_allocator_host_allocator(
 
 static iree_status_t iree_hal_hip_allocator_trim(
     iree_hal_allocator_t* IREE_RESTRICT base_allocator) {
+  // TODO implement
```
```suggestion
  // TODO(awolosyzn): implement hip allocator trim.
```

include either your github ID or an issue number (`TODO(#1234)`) tracking the TODO, and include either a short bit about _what_ is to be done and optionally how (take notes for yourself and future people) - e.g. answer the question "what would trim do here?"

the idea is that we should be able to grep the codebase and roughly see the work to be done in various files - "implement" wouldn't be useful without context.

AWoloszyn replied:
Yep, this was a note to myself to implement this inside the WIP PR.

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -657,16 +709,40 @@ iree_status_t iree_hal_hip_allocator_alloc_async(
     iree_hal_buffer_t* buffer) {
   iree_hal_hip_allocator_t* allocator =
       iree_hal_hip_allocator_cast(base_allocator);
+  if (iree_hal_buffer_allocation_size(buffer) > (size_t)-1 ||
```
use IREE_HOST_SIZE_MAX for this - no need to also compare against size_t

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -657,16 +709,40 @@ iree_status_t iree_hal_hip_allocator_alloc_async(
+      iree_hal_buffer_allocation_size(buffer) > (iree_host_size_t)(-1)) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "buffer allocation too large for hip on this system");
```
when hitting cases like this include the values - for example, `"buffer allocation of %" PRIdsz " too large to map into host address space with max of %" PRIhsz, iree_hal_buffer_allocation_size(buffer), IREE_HOST_SIZE_MAX)`

the goal is to not make the user come in and add breakpoints/printfs/etc - when it's cheap to do so (like here) always include additional information about what specifically caused the error

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -697,8 +773,36 @@ iree_status_t iree_hal_hip_allocator_free_async(
+  }
+  iree_slim_mutex_unlock(&allocator->async_allocation_mutex);
+
+  if (!iree_status_is_ok(status)) {
```
as mentioned in other PRs don't exit early unless you need to:

```
if (iree_status_is_ok(status)) {
  iree_hal_hip_buffer_set_allocation_empty(buffer);
}
return status;
```

there should be very few cases where you need to `if !ok return`

doing this makes it clearer what is a success case vs not, defines a structured place for cleanup (which this may need), and makes inserting tracing easier (as there's a clear place to put the end zone).

in cases where iree_hal_hip_util_tree_insert succeeds but iree_hal_hip_async_allocation_queue_push_back fails you've still got the entry bucket in the tree - is that expected?

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -697,8 +773,36 @@ iree_status_t iree_hal_hip_allocator_free_async(
+    iree_hal_hip_async_allocation_t allocation = {
+        .allocation_size = iree_hal_buffer_allocation_size(buffer),
+        .pointer = device_ptr};
+    iree_hal_hip_async_allocation_queue_push_back(
```
you're ignoring the status here - should be `status = ...`

AWoloszyn committed:
Fixed an off-by-1 in tree.c.

We sometimes accidently freed a pointer that was part of
the cache. I have no idea how the existing tests did not
fail for this.

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

AWoloszyn committed:
Updated from some comments, and implemented trim

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -132,6 +183,30 @@ static iree_allocator_t iree_hal_hip_allocator_host_allocator(
+  }
+  iree_slim_mutex_unlock(&allocator->async_allocation_mutex);
+
   return iree_ok_status();
```
you need to return the status

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -697,9 +798,36 @@ iree_status_t iree_hal_hip_allocator_free_async(
+    iree_hal_hip_buffer_set_allocation_empty(buffer);
+  }
 
   return iree_ok_status();
```
you need to return the status

benvanik reviewed (approved):
two things to fix then lgtm!

AWoloszyn committed:
Returns the statuses and also check the queue count correctly.

We were dropping statuses on the floor, but also
if we had an empty (but existing) allocation queue
we would try to pop the non-existant front thing out of it.

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

AWoloszyn committed:
Moved the async free to the correct timeline.

Have trim wait on all streams so that we know for sure
there are no outstanding allocations.

Also bind the correct context for malloc/free operations.

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

AWoloszyn committed:
Added release callback to asynchronous buffer.

If a buffer was allocated asynchronously, then
we still do actually want to release it properly back
into the asynchronous pool.

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>

benvanik commented on runtime/src/iree/hal/drivers/hip/hip_allocator.c:
```diff
@@ -114,6 +158,27 @@ static void iree_hal_hip_allocator_destroy(
       iree_hal_hip_allocator_cast(base_allocator);
   IREE_TRACE_ZONE_BEGIN(z0);
 
+  iree_slim_mutex_deinitialize(&allocator->async_allocation_mutex);
```
fyi it's generally a good idea to only release a mutex that guards a resource after you clean up the resource (otherwise it's possible functions/callbacks that are called as cleanup may assume it's still valid) - it can also appease tsan (which I wish we ran) as it often doesn't like seeing locked access on thread 2 and then unlocked access on thread 1 (it doesn't know you're deinitializing, only that you aren't using a lock here)

AWoloszyn committed:
Move the mutex deinit to later in the destructor.

Signed-off-by: Andrew Woloszyn <andrew.woloszyn@gmail.com>
